draft-bbs-signatures.md

```diff
@@ -841,6 +845,132 @@ Procedure:
+2. Five (5) integers representing scalars in the range of 1 to r-1 inclusive.
+3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof message commitments.
+
+```
+proof = octets_to_proof(proof_octets)
+
+Inputs:
+
+- proof_octets (REQUIRED), octet string of the form output from
+                           proof_to_octets operation.
+
+Parameters:
+
+- r (REQUIRED), non-negative integer. The prime order of the G_1 and
+                G_2 groups, defined by the ciphersuite.
+- octet_scalar_length (REQUIRED), non-negative integer. The length of
+                                  a scalar octet representation, defined
+                                  by the ciphersuite.
+- octet_point_length (REQUIRED), non-negative integer. The length of
+                                 a point in G1 octet representation,
+                                 defined by the ciphersuite.
+
+Outputs:
+
+- proof, a proof value in the form described above or INVALID
+
+Procedure:
+
+1. proof_len_floor = 3 * octet_point_length + 5 * octet_scalar_length
+
+2. if length(proof_octets) < proof_len_floor, return INVALID
+
+// Points (i.e., (A', Abar, D) in ProofGen) de-serialization.
+3. index = 0
+
+4. for i in (0, 2):
+
+5.     end_index = index + octet_point_length - 1
+
+6.     A_i = octets_to_point_g1(proof_octets[index..end_index])
+
+7.     if A_i is INVALID or Identity_G1, return INVALID
+
+8.     index += octet_point_length
+
+// Scalars (i.e., (c, e^, r2^, r3^, s^, (m^_j1, ..., m^_jU)) in
+// ProofGen) de-serialization.
+9. j = 0
+
+10. while index < length(proof_octets):
+
+11.     end_index = index + octet_scalar_length - 1
+
+12.     s_j = OS2IP(proof_octets[index..end_index])
+
+13.     if s_j = 0 or if s_j >= r, return INVALID
+
+14.     index += octet_scalar_length
+
+15.     j += 1
+
+16. if index is not equal to length(proof_octets), return INVALID
+
+17. msg_commitments = (s_5, ..., s_(k-1))
+
+18. return (A_0, A_1, A_2, s_0, s_1, s_2, s_3, s_4, msg_commitments)
+```
+
+### ProofToOctets
+
+This operation describes how to encode a proof, as computed at step 25 in [ProofGen](#proofgen), to an octet string. The input to the operation MUST be a valid proof.
+
+The inputed proof value must consist of the following components, in that order:
+
+1. Three (3) valid points of the G1 subgroup, different from the identity point of G1 (i.e., `A', Abar, D`, in ProofGen)
+2. Five (5) integers representing scalars in the range of 1 to r-1 inclusive (i.e., `c, e^, r2^, r3^, s^`, in ProofGen).
+3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the un-revealed from the proof messages (i.e., `m^_j1, ..., m^_jU`, in ProofGen, where U the number of un-revealed messages).
+
+```
+proof_octets = proof_to_octets(proof)
+
+Inputs:
+
+- proof (REQUIRED), a BBS proof in the form calculated by ProofGen in
```

Comment:
Wonder if we actually need to supply the arguments destructed here like we do in signature to octets?

Reply:
Since the input has no "fixed" length it may be more complicated to define it ''destructed" (i.e., if we define it as `(A', Abar, D, c, e^, r2^, r3^, s^, (m^_1,...,m^_U))` we will need to define `U` which will be hard here or if we define it like `(A', Abar, D, c, e^, r2^, r3^, s^, msg_commitments`) we mix up structured and destructed forms which misses the point of consistency IMO.

In general, for consistency I would prefer passing the signature in a structured way than passing the proof in a destructed one.

Reply:
Happy to move in that direction do we need to raise an issue to capture that?

Reply:
Updated this PR to make the change for `octets_to_signature` and `signature_to_octets` (+ took the opportunity to make the operations < 72 chars). For other operations the update will most likely happen in #138. Opened #163 to keep track of this.